examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/main.c

```diff
@@ -82,17 +86,280 @@ char * ssid = "SSID";
+
+			xi_connect( in_context_handle,
+					conn_data->username, conn_data->password,
+					conn_data->connection_timeout, conn_data->keepalive_timeout,
+					conn_data->session_type, &on_connected);
+		break;
+		/* connection has been closed */
+		case XI_CONNECTION_STATE_CLOSED:
+			/* unregister task handle */
+			xi_cancel_timed_task( gXivelyTimedTaskHandle );
+			gXivelyTimedTaskHandle = XI_INVALID_TIMED_TASK_HANDLE;
+
+			printf( "[%d] Xively: Connection closed %s:%d reason %d, reconnecting....\r\n",
+					( int ) xi_bsp_time_getcurrenttime_seconds(),
+					conn_data->host, conn_data->port, state );
+
+			xi_connect( in_context_handle,
+					conn_data->username, conn_data->password,
+					conn_data->connection_timeout, conn_data->keepalive_timeout,
+					conn_data->session_type, &on_connected);
+		break;
+		/* connection has been established */
+		case XI_CONNECTION_STATE_OPENED:
+			printf( "[%d] Xively: Connected to %s:%d\r\n",
+					( int ) xi_bsp_time_getcurrenttime_seconds(),
+					conn_data->host, conn_data->port );
+
+			{
+				/* here put the activation code */
+				init_xively_topics( in_context_handle );
+			}
+
+			break;
+		/* something really bad happened */
+		default:
+			printf( "[%d] Xively: Invalid connection status %s:%d\r\n",
+					( int ) xi_bsp_time_getcurrenttime_seconds(),
+					conn_data->host, conn_data->port );
+	};
+}
+
+/* handler for xively mqtt topic subscription */
+void on_mqtt_topic( xi_context_handle_t in_context_handle,
+        xi_sub_call_type_t call_type,
+        const xi_sub_call_params_t* const params,
+        xi_state_t state,
+        void* user_data )
+{
+	/* sanity checks */
+	assert( NULL != user_data );
+
+	/* user data is the mqtt_topic_configuration_t */
+	const mqtt_topic_configuration_t* topic_conf = ( const mqtt_topic_configuration_t* ) user_data;
+
+	switch ( call_type )
+	{
+		case XI_SUB_CALL_SUBACK:
+            if ( XI_MQTT_SUBACK_FAILED == params->suback.suback_status )
+            {
+            	printf( "Subscription failed for %s topic\r\n", params->suback.topic );
+            }
+            else
+            {
+            	printf( "Subscription successful for %s topic\r\n", params->suback.topic );
+            }
+            return;
+        case XI_SUB_CALL_MESSAGE:
+            printf( "received message on %s topic\r\n", params->suback.topic );
+            /* if there is a handler call it with the message payload */
+            if( NULL != topic_conf->on_msg_pull )
+            {
+        		topic_conf->on_msg_pull( params->message.temporary_payload_data,
+        				params->message.temporary_payload_data_length );
+            }
+        	break;
+        default:
+        	return;
+	}
+}
+
+
+void push_mqtt_topics( )
+{
+	int i = 0;
+	for(; i < topics_array_length; ++i )
+	{
+		const mqtt_topic_configuration_t* const topic_config = &topics_array[ i ];
+
+		if( MQTT_TOPIC_PUB == topic_config->topic_descr.topic_type )
+		{
+			if( NULL != topic_config->on_msg_push )

```

Comment:
can't the existence of the on_msg_push define that this topic_config is a PUB type? I mean isn't this representation redundant?